### PR TITLE
[java] ConsecutiveLiteralAppends false positive 

### LINKED ISSUE #1632: [java] ConsecutiveLiteralAppends false positive over catch
**Affects PMD Version:** 6.11.0

**Rule:** [ConsecutiveLiteralAppends](https://pmd.github.io/pmd-6.11.0/pmd_rules_java_performance.html#consecutiveliteralappends)

**Description:** When appending a literal before a try/catch block and also inside the try block, after calling a method, that might throw a exception, PMD falsely reports a `ConsecutiveLiteralAppends` lint.

**Code Sample demonstrating the issue:**

```java
public final class Test {
    public String foo() {
        final StringBuilder sb = new StringBuilder();
        sb.append("literal1");
        try {
            final String res = bar();
            sb.append("literal2").append(res);
        } catch (Exception e) {
            sb.append("literal3");
        }
        return sb.toString();
    }

    public String bar() throws Exception {
        throw new Exception();
    }
}
```

**Running PMD through:** *Gradle*

### PR DESCRIPTION
**Pull request by Shubham-2k17**
**PR Description:**
ASTTryStatement added to BLOCK_PARENTS.
fixes #1632

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 2 configuration errors.<br />
<a href="http://chunk.io/pmd/24baf70e07fb445ca2d1fec35c0a657c/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/ConsecutiveLiteralAppendsRule.java [hunk 1]
```diff
@@ -74,6 +75,7 @@
         BLOCK_PARENTS.add(ASTIfStatement.class);
         BLOCK_PARENTS.add(ASTSwitchStatement.class);
         BLOCK_PARENTS.add(ASTMethodDeclaration.class);
+        BLOCK_PARENTS.add(ASTTryStatement.class);
```
Comment (jsotuyod) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/ConsecutiveLiteralAppendsRule.java]: catch and finally should be added too. A simple test-case having 2 catch blocks do different appends / a catch + finally doing appends would trigger these FP too.

Comment (jsotuyod) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/ConsecutiveLiteralAppendsRule.java]: I'm thinking, we actually may have similar FPs with lambdas, and we should add `ASTLambdaExpression.class` here…

Comment (Shubham-2k17) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/ConsecutiveLiteralAppendsRule.java]: I tried to check lambdas. No FPs.

**Comment by vbrandl** [comment]
I don't know the PMD codebase so I don't know what exactly this PR will change, but I would guess, it would just ignore literals, that are appended inside a try/catch block?

So this won't trigger the rule while it should:
```java
final StringBuilder sb = new StringBuilder();
sb.append("foo");
try {
  sb.append("bar");
  final String res = methodThatMightThrow();
  sb.append(res);
} catch (IOException ioe) {
  ...
}
```

This is only a false positive, if the literal is appended inside the try/catch block, _after_ a throwing method was called. I don't know if this case is covered by the PR.

**Comment by Shubham-2k17** [comment]
Hi @vbrandl,
The PR is currently not covering the case where method throwing exception comes after append in try catch.

**Comment by jsotuyod** [comment]
@Shubham-2k17 @vbrandl interesting! I think you are actually right. `ASTTryStatement` shouldn't be in the block parents, but catch / finally do.

I'm updating, adding that test case along with the lambda ones (which works as already stated), and merging.

**Pull request merged**